File: packages/sdk-typescript/src/logger.ts
/**
 * Logging interface and implementations
 */

export type LogLevel = "debug" | "info" | "warn" | "error" | "none";

export interface Logger {
  debug(message: string, ...args: unknown[]): void;
  info(message: string, ...args: unknown[]): void;
  warn(message: string, ...args: unknown[]): void;
  error(message: string, ...args: unknown[]): void;
}

export interface LoggerConfig {
  level?: LogLevel;
  prefix?: string;
}

/**
 * Console logger implementation with configurable log levels
 */
export class ConsoleLogger implements Logger {
  private level: LogLevel;
  private prefix: string;

  private readonly levels: Record<LogLevel, number> = {
    debug: 0,
    info: 1,
    warn: 2,
    error: 3,
    none: 4,
  };

  constructor(config: LoggerConfig = {}) {
    this.level = config.level ?? "info";
    this.prefix = config.prefix ?? "[Exchange SDK]";
  }

  private shouldLog(level: LogLevel): boolean {
    return this.levels[level] >= this.levels[this.level];
  }

  private formatMessage(level: string, message: string): string {
    return `${this.prefix} ${level.toUpperCase()}: ${message}`;
  }

  debug(message: string, ...args: unknown[]): void {
    if (this.shouldLog("debug")) {
      console.debug(this.formatMessage("debug", message), ...args);
    }
  }

  info(message: string, ...args: unknown[]): void {
    if (this.shouldLog("info")) {
      console.log(this.formatMessage("info", message), ...args);
    }
  }

  warn(message: string, ...args: unknown[]): void {
    if (this.shouldLog("warn")) {
      console.warn(this.formatMessage("warn", message), ...args);
    }
  }

  error(message: string, ...args: unknown[]): void {
    if (this.shouldLog("error")) {
      console.error(this.formatMessage("error", message), ...args);
    }
  }

  setLevel(level: LogLevel): void {
    this.level = level;
  }

  getLevel(): LogLevel {
    return this.level;
  }
}

/**
 * No-op logger that discards all log messages
 */
export class NoopLogger implements Logger {
  debug(): void {}
  info(): void {}
  warn(): void {}
  error(): void {}
}

File: packages/sdk-python/exchange_sdk/logger.py
"""Logging interface and implementations."""

from abc import ABC, abstractmethod
from enum import Enum
from typing import Any


class LogLevel(str, Enum):
    """Log level enumeration."""

    DEBUG = "debug"
    INFO = "info"
    WARN = "warn"
    ERROR = "error"
    NONE = "none"


class Logger(ABC):
    """Abstract logger interface."""

    @abstractmethod
    def debug(self, message: str, *args: Any) -> None:
        """Log debug message."""
        pass

    @abstractmethod
    def info(self, message: str, *args: Any) -> None:
        """Log info message."""
        pass

    @abstractmethod
    def warn(self, message: str, *args: Any) -> None:
        """Log warning message."""
        pass

    @abstractmethod
    def error(self, message: str, *args: Any) -> None:
        """Log error message."""
        pass


class ConsoleLogger(Logger):
    """Console logger implementation with configurable log levels."""

    _LEVELS = {
        LogLevel.DEBUG: 0,
        LogLevel.INFO: 1,
        LogLevel.WARN: 2,
        LogLevel.ERROR: 3,
        LogLevel.NONE: 4,
    }

    def __init__(self, level: LogLevel = LogLevel.INFO, prefix: str = "[Exchange SDK]"):
        """
        Initialize console logger.

        Args:
            level: Minimum log level to display
            prefix: Prefix for log messages
        """
        self.level = level
        self.prefix = prefix

    def _should_log(self, level: LogLevel) -> bool:
        """Check if message should be logged based on level."""
        return self._LEVELS[level] >= self._LEVELS[self.level]

    def _format_message(self, level: str, message: str) -> str:
        """Format log message with prefix and level."""
        return f"{self.prefix} {level.upper()}: {message}"

    def debug(self, message: str, *args: Any) -> None:
        """Log debug message."""
        if self._should_log(LogLevel.DEBUG):
            print(self._format_message("debug", message), *args)

    def info(self, message: str, *args: Any) -> None:
        """Log info message."""
        if self._should_log(LogLevel.INFO):
            print(self._format_message("info", message), *args)

    def warn(self, message: str, *args: Any) -> None:
        """Log warning message."""
        if self._should_log(LogLevel.WARN):
            print(self._format_message("warn", message), *args)

    def error(self, message: str, *args: Any) -> None:
        """Log error message."""
        if self._should_log(LogLevel.ERROR):
            print(self._format_message("error", message), *args)

    def set_level(self, level: LogLevel) -> None:
        """Set log level."""
        self.level = level

    def get_level(self) -> LogLevel:
        """Get current log level."""
        return self.level


class NoopLogger(Logger):
    """No-op logger that discards all log messages."""

    def debug(self, message: str, *args: Any) -> None:
        """Do nothing."""
        pass

    def info(self, message: str, *args: Any) -> None:
        """Do nothing."""
        pass

    def warn(self, message: str, *args: Any) -> None:
        """Do nothing."""
        pass

    def error(self, message: str, *args: Any) -> None:
        """Do nothing."""
        pass
